drivers/linstorvolumemanager.py

```diff
@@ -196,17 +202,88 @@ def _get_controller_uri():
         # Not found, maybe we are trying to create the SR...
         pass
 
-def get_controller_uri():
-    retries = 0
-    while True:
-        uri = _get_controller_uri()
+
+@contextlib.contextmanager
+def shared_reader(path):
+    with open(path, 'r') as f:
+        lock = flock.ReadLock(f.fileno())
+        lock.lock()
+        try:
+            yield f
+        finally:
+            lock.unlock()
+
+
+@contextlib.contextmanager
+def excl_writer(path):
+    with open(path, 'r+') as f:
+        lock = flock.WriteLock(f.fileno())
+        lock.lock()
+        try:
+            yield f
+        finally:
+            lock.unlock()
+
+
+def get_cached_controller_uri(ctx=None):
+    try:
+        with ctx if ctx else shared_reader(CONTROLLER_CACHE_PATH) as f:
+            return f.read().strip()
+    except FileNotFoundError:
+        pass
+    except Exception as e:
+        util.SMlog('Unable to read controller URI cache file at `{}` : {}'.format(CONTROLLER_CACHE_PATH,e))
+
+
+def delete_controller_uri_cache(ctx=None):
+    try:
+        with ctx if ctx else excl_writer(CONTROLLER_CACHE_PATH) as f:
+            f.seek(0)
+            f.truncate()
+    except FileNotFoundError:
+        pass
+    except Exception as e:
+        util.SMlog('Unable to delete uri cache file at `{}` : {}'.format(CONTROLLER_CACHE_PATH, e))
```

Comment:
```suggestion
def delete_controller_uri_cache():
    try:
        with excl_writer(CONTROLLER_CACHE_PATH) as f:
            f.seek(0)
            f.truncate()
    except FileNotFoundError:
        pass
    except Exception as e:
        util.SMlog('Unable to delete URI cache file at `{}`: {}'.format(CONTROLLER_CACHE_PATH, e))
```

Reply:
done in [4f84f23](https://github.com/xcp-ng/sm/pull/83/commits/4f84f2365e9c7ef63fedc9fbdf845b498c73a2ba)